iceoryx_hoofs/test/moduletests/test_relative_pointer.cpp

```diff
@@ -86,72 +96,88 @@ TYPED_TEST_SUITE(RelativePointer_test, Types);
 TYPED_TEST(RelativePointer_test, ConstrTests)
 {
     ::testing::Test::RecordProperty("TEST_ID", "cae7b4d4-86eb-42f6-b938-90a76f01bea5");
-    EXPECT_EQ(BaseRelativePointer::registerPtr(1, this->memoryPartition[0], SHARED_MEMORY_SIZE), true);
-    EXPECT_EQ(BaseRelativePointer::registerPtr(2, this->memoryPartition[1], SHARED_MEMORY_SIZE), true);
-
-    auto ptr0 = this->partitionPtr(0);
-    auto ptr1 = this->partitionPtr(1);
+    EXPECT_EQ(BaseRelativePointer::registerPtr(iox::rp::BaseRelativePointer::id_t{1U},
+                                               static_cast<void*>(this->memoryPartition[0]),
+                                               SHARED_MEMORY_SIZE),
+              true);
+    EXPECT_EQ(BaseRelativePointer::registerPtr(iox::rp::BaseRelativePointer::id_t{2U},
+                                               static_cast<void*>(this->memoryPartition[1]),
+                                               SHARED_MEMORY_SIZE),
+              true);
+
+    auto* ptr0 = this->partitionPtr(0U);
+    auto* ptr1 = this->partitionPtr(1U);
 
     {
-        auto offset = SHARED_MEMORY_SIZE / 2;
-        auto typedPtr = reinterpret_cast<TypeParam*>(ptr0 + offset);
+        auto offset = SHARED_MEMORY_SIZE / 2U;
+        // NOLINTJUSTIFICATION Pointer arithmetic needed for tests
+        // NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic, cppcoreguidelines-pro-type-reinterpret-cast)
+        auto* typedPtr = reinterpret_cast<TypeParam*>(ptr0 + offset);
 
         RelativePointer<TypeParam> rp;
         rp = typedPtr;
         EXPECT_EQ(rp.getOffset(), offset);
-        EXPECT_EQ(rp.getId(), 1);
+        EXPECT_EQ(rp.getId(), 1U);
         EXPECT_NE(rp, nullptr);
     }
 
     {
-        auto typedPtr = reinterpret_cast<TypeParam*>(ptr0);
+        auto* typedPtr = static_cast<TypeParam*>(static_cast<void*>(ptr0));
```

Comment:
It took me some time to distinguish the difference when `reinterpret_cast` is used and when this construct with two `static_cast` is used. Since this is test code, how about just using `auto* typedPtr = reinterpret_cast<TypeParam*>(ptr0); // NOLINT`. It would help with readability and should be sufficient for this specific file. Maybe with a note like
```
// using only NOLINT for reinterpret_cast instead of NOLINTNEXTLINE(cppcoreguidelines-pro-bounds-pointer-arithmetic, cppcoreguidelines-pro-type-reinterpret-cast)
```

Reply:
We are really reinterpreting pointers here so a reinterpret cast is justified. Unlike static_cast it also cannot change the pointer (static_cast does this in some complex inheritance cases). static_cast should not be used here.

static_cast should be used for conversion between primitive types and casting within an inheritance hierarchy only. raw byte buffer casts (source `uint8_t*`, `void*` `char*`) to some other type T* are reinterpret_casts by nature (and requires us to know the actual type which we do here).

Reply:
For me tests are often exemplary code where newbies can learn from. Hence, I would prefer to only use `reinterpret_cast` if really needed. I added a comment above the all the lines that use the cast via `void*` to make the intention clear.

Reply:
Well, `reinterpret_cast<T*>(uPtr)` and `static_cast<T*>(static_cast<void*>(uPtr))` are doing the exact same thing and there is no harm in doing the former one, except a tool telling us not to use it. I'm fine either way.

Personally I'm more annoyed by all this justifications and NOLINTS. Maybe we can just use one big `NOLINTBEGIN` block to disable the warning in cases where it is unavoidable to use a specific construct?

I'm fine to keep it as is for this test.

Reply:
@elBoberido I've not looked at the implementation of either `reinterpret_cast` or `static_cast` but I'm sure they are not doing exactly the same thing, because `static_cast` can be used during compile-time and `reinterpret_cast` can not. It's also likely highly compiler dependent. We can think about writing a wrapper that beforms the trampoline cast.